File: examples/use_after_move_vec.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int main() {
    string origin = "This is a string.";

    vector<std::string> vec;

    // Copy origin and append to vec
    vec.push_back(origin);

    cout << "After pushing origin copy onto vec, origin is unchanged: \"" << origin << "\"\n";

    // Move origin into vec, invalidating origin variable (at runtime)
    vec.push_back(std::move(origin));

    cout << "After move into vec, origin is invalidated: \"" << origin << "\"\n";

    cout << "Contents of vec: \"" << vec[0] << "\", \"" << vec[1] << "\"\n";
}

File: examples/use_after_move_vec.rs
fn main() {
    let origin: String = "This is a string.".into();

    let mut vec = Vec::new();

    // Clone origin and push onto vec
    vec.push(origin.clone());

    println!("After pushing origin clone onto vec, origin is unchanged: {}", origin);

    // Move origin onto vec
    vec.push(origin);

    println!("After moving origin into vec, it cannot be used: {}", origin);

    println!("Contents of vec: {:?}", vec);
}
